### PR TITLE
Implement FE for Test case health

### PR DESCRIPTION
**Pull request by asankov**

### CODE REVIEW

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    })
+}
+
+function drawTable(selector, data) {
+    $(`${selector} > tbody`).remove()
+
+    $(selector).DataTable({
+        data: data,
+        columns: [
+            { data: 'case_summary' },
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: This must be like in search pages: TC-XX: summary and also be a link. Again, smells like utility function (you can do the refactoring later, just make a note of it).

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            { 
+                data: null,
+                render: data => `${data.count.fail} of ${data.count.all}`
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: is there an easy way to add a visual idicator like in the mockups ?

**Comment by asankov** [comment]
> Render all rows in a single table which will make all columns the same width and the page will be more visually pleasing (see screenshot).

This would look quite hacky if we want to have 1 table with multiple header rows between the data columns.

For sure it is not possible out of the box with Datatables. There is a way to hack it, but it would look something like this:

```javascript

    $(selector).DataTable({
        data: [
            ...data.all_fail,
            { header : true},
            ...data.few_fail,
            { header : true},
            ...data.flaky
        ],
        columns: [
            {
                data: null,
                render: data => data.header ? '' : `<a href=""> TC-${data.case_id} </a> - ${data.case_summary}`

            },
            {
                data: null,
                render: data => data.header ? '' : `${data.count.fail} of ${data.count.all}`
            },
            {
                data: null,
                render: data => data.header ? '' :  `${data.count.fail / data.count.all * 100}`
            },
        ],
        createdRow: (row, data, dataIndex) => {
            if (data.header) {
                // apply the classes to row, so that it looks like a header
            }
        },
        paging: false,
        ordering: false,
        dom: "t"
    });
```
in this example we pass to Datatables an array of different objects:
some of them are data object, they have `case_summary`, `case_id`, `count`, etc.
the others are `{ header: true}` - this is to signal to Datatables when it should draw a header column, see `createdRow` - this property allows us to manipulate the already drawn rows. There we can do everything that we can do with `jQuery` elsewhere, add text to DOM element, add classes, add css etc.

I am still having troubles managing to apply the needed classes, so that an ordinary row for Datables looks like a header, but I guess I need more time to find a work around of that, or just find the needed styles.

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js (hunk 1 above)
Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: #1127

**Comment by asankov** [comment]
> Render all rows in a single table which will make all columns the same width and the page will be more visually pleasing (see screenshot).

I managed to hack DataTables in order to achieve this.

> Most Failed Tests should probably be 'Most frequently failing', i.e. use labels which are more natural and explain what the data means

Done


> Don't capitalize titles in the table headers

Done

> Inspect your status calculation logic b/c I see some bogus results. For example "Merchandise: T-shirts" has 0% failures, which should mean it is always stable, same for the table at the bottom. For reference see status matrix for the same Product and TP.

Done

File: tcms/telemetry/api.py [hunk 3]
```diff
@@ -240,6 +261,34 @@ def _build_result_from_data(data):
     return result
 
 
+def _build_result_from_data_grouped(data):
```
Comment (atodorov) [inline tcms/telemetry/api.py]: Unused function

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js (hunk 2 above)
Comment (atodorov) [reply tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: ping ^^^

File: tcms/telemetry/templates/telemetry/testing/test-case-health.html [hunk 4]
```diff
@@ -0,0 +1,100 @@
+{% extends "base.html" %}
+{% load i18n %}
+{% load static %}
+
+{% block title %}{% trans "Testing Status Matrix" %}{% endblock %}
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/test-case-health.html]: wrong title

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            { text: "Flaky tests" },
+            { header: true },
+            ...data.flaky,
+            { text: "Almost stable" },
+            { header: true },
+            ...data.almost_stable,
+        ]);
+
+        // drawTable('#table-few-failed', data.few_fail);
+        // drawTable('#table-flaky', data.flaky);
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: commented out code

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+function renderTestCaseColumn(data) {
+    if (data.header) {
+        return "";
+    }
+    if (data.text) {
+        return data.text;
+    }
+    return `<a href="/case/${data.case_id}"> TC-${data.case_id} </a>: ${data.case_summary}`;
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: No space between `<a>` tags

File: tcms/telemetry/templates/telemetry/testing/test-case-health.html [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+                    {% include "datetimepicker_script.html" with selector="#id_before" %}
+                </div>
+            </div>
+        </form>
+
+        <table class="table table-bordered table-with-horizontal-scroll cell-border" id="table-most-failed">
+            <thead>
+                <tr>
+                    <th class="header" colspan="3"> {% trans "Most Failed Tests" %} </th>
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/test-case-health.html]: `Most Failed Tests` -> `Always failing`

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        return "";
+    }
+    return `${data.count.fail} of ${data.count.all}`;
+}
+
+function renderPercentColumn(data) {
+    if (data.header || data.text) {
+        return "";
+    }
+    return `${data.count.fail / data.count.all * 100}`;
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: Round up to 1 decimal point. I get things like:
```
54.54545454545454
9.090909090909092
```

File: tcms/telemetry/templates/telemetry/testing/test-case-health.html [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+                </tr>
+                <tr>
+                    <th class="header"> {% trans "Test Case" %}</th>
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/test-case-health.html]: Instead of `Test Case` is it possible to either:

1) have these headers only at the top or
2) have the category name in this cell and show an empty colspan=3 row between them ?

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        return data.text;
+    }
+    return `<a href="/case/${data.case_id}"> TC-${data.case_id} </a>: ${data.case_summary}`;
+}
+
+function renderFailedExecutionsColumn(data) {
+    if (data.header || data.text) {
+        return "";
+    }
+    return `${data.count.fail} of ${data.count.all}`;
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: How about a progress bar, similar to the one in the screenshots? Is there anything in Patternfly which can be used quickly ?

File: tcms/telemetry/templates/telemetry/testing/test-case-health.html (hunk 9 above)
Comment (asankov) [reply tcms/telemetry/templates/telemetry/testing/test-case-health.html]: 2. I don't get that. Seems the same as 1.

1. Sure. It even looks better this way IMO :
image

Do you want to leave it this way?

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js (hunk 2 above)
Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: As you can see DataTables is pretty limited in this aspect. I could always hack it to insert another row and make it looks like something else, but I think that the way to go here is with 3 different tables, with synchronized column width (I am sure this could be done with css), so that we have the freedom with the layout between them, so that may be a good point for refactoring and improvements.

Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: Sorry, I just realized we are talking about completely different things here:

we could use: 
https://www.patternfly.org/v3/pattern-library/data-visualization/utilization-bar-chart/#overview or https://www.patternfly.org/v3/pattern-library/data-visualization/bullet-chart/

but IMO this should be an improvement and not part of the scope of this PR

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js (hunk 10 above)
Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: see my other comment

Comment (atodorov) [reply tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: I don't see a comment about progress bar widget so put a link to it.

File: tcms/telemetry/api.py [hunk 11]
```diff
@@ -171,9 +171,30 @@ def _append_status_counts_to_result(count, result):
...
 
+    for element in data:
+        failed_percent = element['count']['fail'] / element['count']['all']
+        if failed_percent == 1:
+            result['all_fail'].append(element)
+        elif failed_percent >= 0.5:
+            result['flaky'].append(element)
+        elif failed_percent < 0.5:
+            result['almost_stable'].append(element)
+
```
Comment (atodorov) [inline tcms/telemetry/api.py]: Sort elements in the lists before returning.

Comment (atodorov) [reply tcms/telemetry/api.py]: Note: I will re-think the band distribution which will necessitate update to the numeric constants used in this function and also in `_build_result_from_data()`.

I'm wondering if there's a possibility to have these ranges defined somewhere and use them. How about tuples (lower, upper) ?

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js (hunk 10 above)
Comment (asankov) [reply tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: https://github.com/kiwitcms/Kiwi/pull/1096#discussion_r324798497

this one

File: tcms/telemetry/api.py (hunk 11 above)
Comment (asankov) [reply tcms/telemetry/api.py]: here, the elements are already sorted. e.g. `data` is sorted, because it comes from `test_case_health` and `test_case_health` is sorting before returning:
```python
def test_case_health(query=None):
    # more code

    data.sort(key=_sort_by_passing_rate)

    return _build_result_from_data(data)
```
or do you mean sort by something else?

Comment (asankov) [reply tcms/telemetry/api.py]: > I'm wondering if there's a possibility to have these ranges defined somewhere and use them. 

Sure, we can define them whenever we decide. config files, env variables, you name it. 

> How about tuples (lower, upper) ?

I don't understand that part of the comment

File: tcms/telemetry/static/telemetry/js/testing/test-case-health.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+                render: renderPercentColumn
+            },
+        ],
+        paging: false,
+        ordering: false,
+        dom: "t"
+    });
+}
+
+function renderTestCaseColumn(data) {
```
Comment (asankov) [inline tcms/telemetry/static/telemetry/js/testing/test-case-health.js]: this could be extracted to `telemetry/utils.js`, but let's do that in a separate PR, which will include removing the duplicated logic as well
